File: day2.py
""" Solution to Day 2 """

import os

from intcode import Computer


def _part1(memory):
    computer = Computer(memory)
    memory = computer.run(12, 2)
    print("part1:", memory[0])


def _part2(memory, target):
    computer = Computer(memory)
    for noun in range(100):
        for verb in range(100):
            memory = computer.run(noun, verb)
            if memory[0] == target:
                print("part2:", noun*100 + verb)
                return

    raise ValueError("Unable to find a valid noun/verb")


def _main():
    with open(os.path.join("inputs", "day2.txt")) as file:
        memory = [int(value) for value in file.read().split(',')]

    _part1(memory)
    _part2(memory, 19690720)


if __name__ == "__main__":
    _main()

File: day2.hs
import Intcode
import Control.Applicative
import System.IO
import System.FilePath
import Debug.Trace

makeTuples = liftA2 (,)

testNounVerb :: [Integer] -> (Integer, Integer) -> (Integer, (Integer, Integer))
testNounVerb program nounVerb = do
    let (noun, verb) = nounVerb
    let a:_:_:program' = program
    let nvProgram = a:noun:verb:program'
    let result = head (Intcode.runProgram nvProgram)
    (result, nounVerb)

part1 :: [Integer] -> Integer
part1 program = do
    let (result, _) = testNounVerb program (12, 2)
    result

part2 :: [Integer] -> Integer -> Integer
part2 program target = do
    let nounVerbs = makeTuples [0..99] [0..99]
    let results = map (testNounVerb program) nounVerbs
    let (_, (noun, verb)):_ = filter (\(result, _) -> result == target) results
    noun*100 + verb

prep :: Char -> Char
prep char
    | char == ',' = ' '
    | otherwise = char


main = do
    let path = "inputs" ++ [pathSeparator] ++ "day2.txt"
    contents <- readFile path
    let values = words (map prep contents)
        program = map read values :: [Integer]

    putStrLn ("part 1: " ++ (show (part1 program)))
    putStrLn ("part 2: " ++ (show (part2 program 19690720)))
